src/main/java/headcube/Parser.java

```diff
@@ -0,0 +1,86 @@
+package headcube;
+
+/**
+ * The Parser class handles the parsing of user input and executing corresponding actions
+ * in the HeadCube application.
+ */
+public class Parser {
+    private Ui ui;
+    private TaskList taskList;
+    private Storage storage;
+
+    /**
+     * Constructor of a Parser object with references to the UI, TaskList, and Storage.
+     *
+     * @param ui The UI for displaying messages to the user.
+     * @param taskList The list of tasks to manage.
+     * @param storage The storage to save and load tasks.
+     */
+    public Parser(Ui ui, TaskList taskList, Storage storage) {
+        this.ui = ui;
+        this.taskList = taskList;
+        this.storage = storage;
+    }
+
+    /**
+     * Parses the given input string and performs the corresponding action.
+     *
+     * @param input The user input to parse.
+     * @return The response to the input
+     */
+    public String parse(String input) {
+        try {
+            String[] split = input.split(" ", 2);
+
+            if (input.equals("list")) {
+                return ui.list(taskList);
+            } else if (split[0].equals("mark")) {
+                return taskList.mark(Integer.parseInt(split[1]));
```

Comment:
Consider handling [NumberFormatException](https://docs.oracle.com/javase/8/docs/api/java/lang/NumberFormatException.html) - if the string does not contain a parsable integer. There are other cases of this below too.

Reply:
Also consider the case where the integer is not a valid task index, e.g. -1 or 9999999